src/simulation/dynamics/facetSRPDynamicEffector/facetSRPDynamicEffector.cpp

```diff
@@ -0,0 +1,176 @@
+}
+
+/*! The destructor. */
+FacetSRPDynamicEffector::~FacetSRPDynamicEffector()
+{
+}
+
+/*! The reset member function. This method checks to ensure the input message is linked.
+ @return void
+ @param CurrentSimNanos [ns]  Time the method is called
+*/
+void FacetSRPDynamicEffector::Reset(uint64_t CurrentSimNanos)
+{
+    // Check if input message has not been included
+	if (!this->sunInMsg.isLinked()) {
+		bskLogger.bskLog(BSK_ERROR, "FacetSRPDynamicEffector.sunInMsg was not linked.");
+	}
+}
+
+/*! The SRP dynamic effector does not write any output messages.
+ @return void
+ @param CurrentClock [ns] Time the method is called
+*/
+void FacetSRPDynamicEffector::WriteOutputMessages(uint64_t CurrentClock)
+{
+}
+
+/*! This member function populates the spacecraft geometry structure with user-input facet information.
+ @return void
+ @param area  [m^2] Facet area
+ @param specCoeff  Facet spectral reflection optical coefficient
+ @param diffCoeff  Facet diffuse reflection optical coefficient
+ @param normal_B  Facet normal expressed in B frame components
+ @param locationPntB_B  [m] Facet location wrt point B in B frame components
+*/
+void FacetSRPDynamicEffector::addFacet(double area, double specCoeff, double diffCoeff, Eigen::Vector3d normal_B, Eigen::Vector3d locationPntB_B)
+{
+    this->scGeometry.facetAreas.push_back(area);
+    this->scGeometry.facetSpecCoeffs.push_back(specCoeff);
+    this->scGeometry.facetDiffCoeffs.push_back(diffCoeff);
+    this->scGeometry.facetNormals_B.push_back(normal_B);
+    this->scGeometry.facetLocationsPntB_B.push_back(locationPntB_B);
+    this->numFacets = this->numFacets + 1;
+}
+
+/*! This method is used to link the faceted SRP effector to the hub attitude and position,
+which are required for calculating SRP forces and torques.
+ @return void
+ @param states  Dynamic parameter states
+*/
+void FacetSRPDynamicEffector::linkInStates(DynParamManager& states)
+{
+    this->hubSigma = states.getStateObject("hubSigma");
+    this->hubPosition = states.getStateObject("hubPosition");
+}
+
+/*! This method computes the body forces and torques for the SRP effector.
+ @return void
+ @param integTime  [s] Time the method is called
+ @param timeStep  [s] Simulation time step
+*/
+void FacetSRPDynamicEffector::computeForceTorque(double integTime, double timeStep)
+{
+    // Read the input message
+    SpicePlanetStateMsgPayload sunMsgBuffer;
+    sunMsgBuffer = sunInMsg.zeroMsgPayload;
+    sunMsgBuffer = this->sunInMsg();
+
+    // Calculate the Sun position with respect to the inertial frame, expressed in inertial frame components
+    this->r_SN_N = cArray2EigenVector3d(sunMsgBuffer.PositionVector);
+
+    // Compute dcm_BN using MRP transformation
+    Eigen::MRPd sigmaBN;
+    sigmaBN = (Eigen::Vector3d)this->hubSigma->getState();
+    Eigen::Matrix3d dcm_BN = sigmaBN.toRotationMatrix().transpose();
+
+    // Store the hub B frame position with respect to the inertial frame
+    Eigen::Vector3d r_BN_N;
```

Comment:
:small_red_triangle_down: You can define and instantiate many of these variables in line. I recommend doing that. E.g. `Eigen::Vector3d r_BN_N = this->hubPosition->getState();`

Reply:
Done. I needed to keep the lines separate for the MRP because it would not build if I used a single line.

Reply:
True. That's a deficiency in the MRPd code we've written. There are a number of improvements we should make there.